applications/FluidDynamicsApplication/custom_utilities/cfd_utilities.cpp

```diff
@@ -0,0 +1,429 @@
+
+double CalculateLinearLogarithmicWallFunctionBasedYPlusAndUtau(
+    array_1d<double, 3>& rFrictionVelocity,
+    const array_1d<double, 3>& rWallVelocity,
+    const array_1d<double, 3>& rNormal,
+    const double KinematicViscosity,
+    const double WallHeight,
+    const double VonKarman,
+    const double WallSmoothness,
+    const int MaxIterations,
+    const double Tolerance)
+{
+    KRATOS_TRY
+
+    // calculating unit normal since rNormal contains surface area of the condition
+    const double surface_area = norm_2(rNormal);
+    const array_1d<double, 3>& unit_normal = rNormal / surface_area;
+
+    // calculate tangential velocity
+    noalias(rFrictionVelocity) =
+        rWallVelocity - unit_normal * inner_prod(rWallVelocity, unit_normal);
+    const double wall_velocity = norm_2(rFrictionVelocity);
+
+    // calculate linear log region limit
+    const double limit_y_plus = CalculateLinearLogarithmicWallFunctionBasedYPlusLimit(
+        VonKarman, WallSmoothness, MaxIterations, Tolerance);
+
+    // linear region
+    double u_tau = std::sqrt(wall_velocity * KinematicViscosity / WallHeight);
+    double y_plus = u_tau * WallHeight / KinematicViscosity;
+    const double inv_kappa = 1.0 / VonKarman;
+
+    // log region
+    if (y_plus > limit_y_plus)
+    {
+        int iter = 0;
+        double dx = 1e10;
+        double u_plus = inv_kappa * std::log(y_plus) + WallSmoothness;
+
+        while (iter < MaxIterations && std::fabs(dx) > Tolerance * u_tau)
+        {
+            // Newton-Raphson iteration
+            double f = u_tau * u_plus - wall_velocity;
+            double df = u_plus + inv_kappa;
+            dx = f / df;
+
+            // Update variables
+            u_tau -= dx;
+            y_plus = WallHeight * u_tau / KinematicViscosity;
+            u_plus = inv_kappa * std::log(y_plus) + WallSmoothness;
+            ++iter;
+        }
+
+        KRATOS_WARNING_IF("CFDUtilities", iter >= MaxIterations)
+            << "CalculateLinearLogarithmicWallFunctionBasedYPlusAndUtau "
+               "couldn't converge Newton-Raphson. Residual is "
+            << dx << std::endl;
+    }
+
+    noalias(rFrictionVelocity) =
+        rFrictionVelocity *
+        (-1.0 * u_tau /
+         (wall_velocity <= std::numeric_limits<double>::epsilon() ? 1.0 : wall_velocity));
+
+    return y_plus;
+
+    KRATOS_CATCH("");
+}
+
+void CalculateYPlusAndUTauForConditions(
+    ModelPart& rModelPart,
+    const Variable<double>& rKinematicViscosityVariable,
+    const std::function<double(
+        array_1d<double, 3>&, const GeometryType&, const array_1d<double, 3>&, const Vector&, const double, const double, const double)>& rYPlusAndUTauCalculationMethod)
+{
+    KRATOS_TRY
+
+    const int domain_size = rModelPart.GetProcessInfo()[DOMAIN_SIZE];
+    const std::function<void(array_1d<double, 3>&, const ConditionType&)> normal_calculation_method =
+        (domain_size == 2) ? CalculateConditionNormal<2> : CalculateConditionNormal<3>;
```

Comment:
I think you can avoid this complexity by using the condition normal.

Reply:
done :)